File: app/src/main/java/com/xfl/kakaotalkbot/DebugModeScreen.java
package com.xfl.kakaotalkbot;

import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.Toast;

import com.xfl.kakaotalkbot.MessageUI.MessageListAdapter;
import com.xfl.kakaotalkbot.MessageUI.UserMessage;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


/**
 * Created by XFL on 2/19/2018.
 */

public class DebugModeScreen extends AppCompatActivity {

    static Map<String, List<UserMessage>> savedMessageList = new HashMap<>();
    static List<UserMessage> messageList = new ArrayList<>();
    static RecyclerView mMessageRecycler;
    static MessageListAdapter mMessageAdapter;
    String scriptName;

    public static void appendReply(final String value) {

        //messageList.add(new UserMessage("BOT", value));

        NotificationListener.UIHandler.post(new Runnable() {
            @Override
            public void run() {
                mMessageAdapter.addItem(new UserMessage(true, value, "BOT"), mMessageAdapter.getItemCount());
                //mMessageAdapter.notifyItemInserted(mMessageAdapter.getItemCount() - 1);
                mMessageRecycler.scrollToPosition(messageList.size() - 1);
            }
        });


    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.fragment_debugmode);
        final Button sendBtn = findViewById(R.id.button_chatbox_send);
        final EditText room = findViewById(R.id.txt_room);
        final EditText sender = findViewById(R.id.txt_sender);
        final EditText msgTxt = findViewById(R.id.edittext_chatbox);
        final CheckBox chk_groupchat = findViewById(R.id.chk_groupchat);
        getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);
        scriptName = getIntent().getExtras().getString("scriptName");
        chk_groupchat.setChecked(MainApplication.getContext().getSharedPreferences("debugGroupChat", 0).getBoolean(scriptName, false));
        room.setText(MainApplication.getContext().getSharedPreferences("debugRoom", 0).getString(scriptName, ""));
        sender.setText(MainApplication.getContext().getSharedPreferences("debugSender", 0).getString(scriptName, ""));
        if (savedMessageList.get(scriptName) == null)
            savedMessageList.put(scriptName, new ArrayList<UserMessage>());
        messageList = savedMessageList.get(scriptName);

        mMessageRecycler = findViewById(R.id.recyclerview_message_list);
        mMessageAdapter = new MessageListAdapter(this, messageList);
        mMessageRecycler.setLayoutManager(new LinearLayoutManager(DebugModeScreen.this));
        mMessageRecycler.setAdapter(mMessageAdapter);

        sendBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                if (TextUtils.isEmpty(msgTxt.getText().toString())) return;
                if (NotificationListener.container.get(scriptName) == null) {
                    Toast.makeText(DebugModeScreen.this, DebugModeScreen.this.getResources().getString(R.string.please_compile_first), Toast.LENGTH_SHORT).show();
                    return;
                }
                if (room.getText().toString().isEmpty()) {
                    room.setText("DEBUGROOM");
                }
                if (sender.getText().toString().isEmpty()) {
                    sender.setText("DEBUGSENDER");
                }
                MainApplication.getContext().getSharedPreferences("debugGroupChat", 0).edit().putBoolean(scriptName, chk_groupchat.isChecked()).apply();
                MainApplication.getContext().getSharedPreferences("debugSender", 0).edit().putString(scriptName, sender.getText().toString()).apply();
                MainApplication.getContext().getSharedPreferences("debugRoom", 0).edit().putString(scriptName, room.getText().toString()).apply();
                mMessageAdapter.addItem(new UserMessage(false, msgTxt.getText().toString(), sender.getText().toString()), mMessageAdapter.getItemCount());
                //messageList.add(new UserMessage("USER", msgTxt.getText().toString()));
                int newMsgPosition = mMessageAdapter.getItemCount() - 1;
                //mMessageAdapter.notifyItemInserted(newMsgPosition);
                mMessageRecycler.scrollToPosition(newMsgPosition);
                NotificationListener.debugRoom = room.getText().toString();

                Thread thr = new Thread(new Runnable() {
                    @Override
                    public void run() {
                        NotificationListener.callResponder(scriptName, room.getText().toString(), msgTxt.getText().toString(), sender.getText().toString(), chk_groupchat.isChecked(), new ImageDB(BitmapFactory.decodeResource(MainApplication.getContext().getResources(), R.drawable.ic_lock_bugreport)), "DEBUGMODE", null, true);

                    }
                });
                thr.start();


                msgTxt.setText("");
            }
        });


    }

    /*private void refreshView() {
        mMessageAdapter.notifyItemInserted(messageList.size() - 1);
    }*/

    @Override
    protected void onDestroy() {
        super.onDestroy();
        savedMessageList.put(scriptName, messageList);

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.clear_log, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_clearlogs) {
            mMessageAdapter.clear();
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/xfl/kakaotalkbot/DebugModeScreen.kt
package com.xfl.kakaotalkbot

import android.graphics.BitmapFactory
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.WindowManager
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import android.widget.Toast
import com.xfl.kakaotalkbot.MessageUI.MessageListAdapter
import com.xfl.kakaotalkbot.MessageUI.UserMessage
import java.util.*


/**
 * Created by XFL on 2/19/2018.
 */

class DebugModeScreen : AppCompatActivity() {
    internal var scriptName: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.fragment_debugmode)
        val sendBtn = findViewById<Button>(R.id.button_chatbox_send)
        val room = findViewById<EditText>(R.id.txt_room)
        val sender = findViewById<EditText>(R.id.txt_sender)
        val msgTxt = findViewById<EditText>(R.id.edittext_chatbox)
        val chk_groupchat = findViewById<CheckBox>(R.id.chk_groupchat)

        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE)
        scriptName = intent.extras!!.getString("scriptName")
        chk_groupchat.isChecked = MainApplication.context!!.getSharedPreferences("debugGroupChat", 0).getBoolean(scriptName, false)
        room.setText(MainApplication.context!!.getSharedPreferences("debugRoom", 0).getString(scriptName, ""))
        sender.setText(MainApplication.context!!.getSharedPreferences("debugSender", 0).getString(scriptName, ""))
        if (room.text.isEmpty()) {
            room.requestFocus()
        } else if (sender.text.isEmpty()) {
            sender.requestFocus()
        } else {
            msgTxt.requestFocus()
        }
        if (savedMessageList[scriptName] == null)
            savedMessageList[scriptName] = ArrayList()
        messageList = savedMessageList[scriptName]

        mMessageRecycler = findViewById(R.id.recyclerview_message_list)
        mMessageAdapter = MessageListAdapter(this, messageList)
        mMessageRecycler.layoutManager = LinearLayoutManager(this@DebugModeScreen)
        mMessageRecycler.adapter = mMessageAdapter

        sendBtn.setOnClickListener(View.OnClickListener {
            if (TextUtils.isEmpty(msgTxt.text.toString())) return@OnClickListener
            if (ScriptsManager.container[scriptName!!] == null) {
                Toast.makeText(this@DebugModeScreen, this@DebugModeScreen.resources.getString(R.string.please_compile_first), Toast.LENGTH_SHORT).show()
                return@OnClickListener
            }
            if (room.text.toString().isEmpty()) {
                room.setText("DEBUGROOM")
            }
            if (sender.text.toString().isEmpty()) {
                sender.setText("DEBUGSENDER")
            }
            MainApplication.context!!.getSharedPreferences("debugGroupChat", 0).edit().putBoolean(scriptName, chk_groupchat.isChecked).apply()
            MainApplication.context!!.getSharedPreferences("debugSender", 0).edit().putString(scriptName, sender.text.toString()).apply()
            MainApplication.context!!.getSharedPreferences("debugRoom", 0).edit().putString(scriptName, room.text.toString()).apply()

            messageList?.add(UserMessage(false, msgTxt.text.toString(), sender.text.toString()));
            val newMsgPosition = mMessageAdapter.itemCount - 1
            //mMessageAdapter.notifyItemInserted(newMsgPosition);
            mMessageRecycler.scrollToPosition(newMsgPosition)
            NotificationListener.debugRoom = room.text.toString()
            val scname = scriptName
            //if(NotificationListener.threads[scname]==null)NotificationListener.threads[scname!!]=ArrayList<Thread?>()
            // val id=NotificationListener.threads[scname]!!.size
            val thr = Thread(Runnable {
                NotificationListener.callResponder(scriptName!!, room.text.toString(), msgTxt.text.toString(), sender.text.toString(), chk_groupchat.isChecked, ImageDB(BitmapFactory.decodeResource(MainApplication.context!!.resources, R.drawable.ic_lock_bugreport)), "DEBUGMODE",/*id,*/ null, true)
                //   NotificationListener.threads[scname]!![id]=null
            })

            //   NotificationListener.threads[scname]!!.add(thr)
            thr.start()


            msgTxt.setText("")
        })


    }

    /*private void refreshView() {
        mMessageAdapter.notifyItemInserted(messageList.size() - 1);
    }*/

    override fun onDestroy() {
        super.onDestroy()
        savedMessageList[scriptName] = messageList

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.clear_log, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.action_clearlogs) {
            mMessageAdapter.clear()
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {

        internal var savedMessageList: MutableMap<String?, MutableList<UserMessage>?> = HashMap()
        internal var messageList: MutableList<UserMessage>? = ArrayList()
        lateinit internal var mMessageRecycler: RecyclerView
        lateinit internal var mMessageAdapter: MessageListAdapter

        fun appendReply(value: String?) {

            //messageList.add(new UserMessage("BOT", value));

            NotificationListener.UIHandler!!.post {
                messageList?.add(UserMessage(true, value, "BOT"))
                mMessageAdapter.notifyItemInserted(mMessageAdapter.getItemCount() - 1);
                mMessageRecycler.scrollToPosition(messageList!!.size - 1)
            }


        }
    }
}
